File: src/sql/postgres/protocol/NotificationResponse.zig
const NotificationResponse = @This();

pid: int4 = 0,
channel: bun.ByteList = .{},
payload: bun.ByteList = .{},

pub fn deinit(this: *@This()) void {
    this.channel.clearAndFree(bun.default_allocator);
    this.payload.clearAndFree(bun.default_allocator);
}

pub fn decodeInternal(this: *@This(), comptime Container: type, reader: NewReader(Container)) !void {
    const length = try reader.length();
    bun.assert(length >= 4);

    this.* = .{
        .pid = try reader.int4(),
        .channel = (try reader.readZ()).toOwned(),
        .payload = (try reader.readZ()).toOwned(),
    };
}

pub const decode = DecoderWrap(NotificationResponse, decodeInternal).decode;

const bun = @import("bun");
const DecoderWrap = @import("./DecoderWrap.zig").DecoderWrap;
const NewReader = @import("./NewReader.zig").NewReader;

const types = @import("../PostgresTypes.zig");
const int4 = types.int4;

File: src/sql/postgres/protocol/NotificationResponse.rs
use super::new_reader::NewReader;
use crate::postgres::AnyPostgresError;
use crate::shared::Data;

pub struct NotificationResponse {
    pub channel: Data,
    pub payload: Data,
}

impl NotificationResponse {
    pub fn decode_internal<Container: super::new_reader::ReaderContext>(
        mut reader: NewReader<Container>,
    ) -> Result<Self, AnyPostgresError> {
        let mut remaining = reader.body_length()?;
        if remaining < 4 {
            return Err(AnyPostgresError::InvalidMessage);
        }
        reader.int4()?;
        remaining -= 4;
        let (channel, consumed) = reader.string_within(remaining)?;
        remaining -= consumed;
        let (payload, _) = reader.string_within(remaining)?;

        Ok(Self { channel, payload })
    }
}
